src/scanner.c

```diff
@@ -187,10 +195,42 @@ ts_scanner_start_scan(ScannerCtx *ctx, InternalScannerCtx *ictx)
 
 	ictx->sctx = ctx;
 	ictx->closed = false;
+	ictx->registered_snapshot = false;
 
 	scanner = scanner_ctx_get_scanner(ctx);
 
 	scanner->openscan(ictx);
+
+	if (ctx->snapshot == NULL)
+	{
+		/*
+		 * We use SnapshotSelf by default, for historical reasons mostly, but
+		 * we probably want to move to an MVCC snapshot as the default. The
+		 * difference is that a Self snapshot is an "instant" snapshot and can
+		 * see its own changes. More importantly, however, unlike an MVCC
+		 * snapshot, a Self snapshot is not subject to the strictness of
+		 * SERIALIZABLE isolation mode.
+		 *
+		 * This is important in case of, e.g., concurrent chunk creation by
+		 * two transactions; we'd like a transaction to use a new chunk as
+		 * soon as the creating transaction commits, so that there aren't
+		 * multiple transactions creating the same chunk and all but one fails
+		 * with a conflict. However, under SERIALIZABLE mode a transaction is
+		 * only allowed to read data from transactions that were committed
+		 * prior to transaction start. This means that two or more
+		 * transactions that create the same chunk must have all but the first
+		 * committed transaction fail.
+		 *
+		 * Therefore, we probably want to exempt internal bookkeeping metadata
+		 * from full SERIALIZABLE semantics (at least in the case of chunk
+		 * creation), or otherwise the INSERT behavior will be different for
+		 * hypertables compared to regular tables under SERIALIZABLE
+		 * mode.
+		 */
+		ctx->snapshot = RegisterSnapshot(GetSnapshotData(SnapshotSelf));
+		ictx->registered_snapshot = true;
```

Comment:
It cannot be NULL because a snapshot is always used and this the pointer will always be non-NULL.

The boolean is needed since there are two ways a snapshot is set: 1. by the user, in which case no registration should occur here, or 2. set automatically by the scanner when no user snapshot is specified. Thus, this boolean is here to separate between these two cases, and specifically to unregister the snapshot for case 2.